packages/hebao_v1/contracts/stores/GuardianStore.sol

```diff
@@ -0,0 +1,262 @@
+        )
+        public
+        view
+        returns (Data.Guardian[] memory _guardians)
+    {
+        Wallet storage w = wallets[wallet];
+        _guardians = new Data.Guardian[](w.guardians.length);
+        uint index = 0;
+        for (uint i = 0; i < w.guardians.length; i++) {
+            Data.Guardian memory g = w.guardians[i];
+            if (_isActivatedOrPendingActivation(g,includePendingActive)) {
+                _guardians[index] = g;
+                index++;
+            }
+        }
+        assembly { mstore(_guardians, index) }
+    }
+
+    function numGuardians(
+        address wallet,
+        bool    includePendingActive
+        )
+        public
+        view
+        returns (uint count)
+    {
+        Wallet storage w = wallets[wallet];
+        for (uint i = 0; i < w.guardians.length; i++) {
+            Data.Guardian memory g = w.guardians[i];
+            if (_isActivatedOrPendingActivation(g, includePendingActive)) {
+                count++;
+            }
+        }
+    }
+
+    function removeAllGuardians(address wallet)
+        public
+        onlyWalletModule(wallet)
+    {
+        Wallet storage w = wallets[wallet];
+        for (uint i = 0; i < w.guardians.length; i++) {
+            delete w.guardianIdx[w.guardians[i].addr];
+        }
+        delete w.guardians;
+    }
+
+    function cancelPendingGuardians(address wallet)
+        public
+        onlyWalletModule(wallet)
+    {
+        Wallet storage w = wallets[wallet];
+        for (uint i = 0; i < w.guardians.length; i++) {
+            Data.Guardian memory g = w.guardians[i];
+            if (_isPendingExpire(g)) {
+                w.guardians[i].validUntil = 0;
+            }
+            if (_isPendingActivation(g)) {
+                w.guardians[i].validSince = 0;
+            }
+        }
+        _cleanExpiredGuardians(wallet);
+    }
+
+    function addGuardian(
+        address wallet,
+        address addr,
+        uint    validSince
+        )
+        public
+        onlyWalletModule(wallet)
+    {
+        require(validSince > 0, "INVALID_VALID_SINCE");
+        require(addr != address(0), "ZERO_ADDRESS");
+
+        Wallet storage w = wallets[wallet];
+        uint pos = w.guardianIdx[addr];
+
+        if(pos == 0) {
+            // Add the new guardian
+            Data.Guardian memory g = Data.Guardian(
+                addr,
+                validSince.toUint40(),
+                uint40(0)
+            );
+            w.guardians.push(g);
+            w.guardianIdx[addr] = w.guardians.length;
+
+            _cleanExpiredGuardians(wallet);
+        } else {
+            Data.Guardian memory g = w.guardians[pos - 1];
+            bool isGuardianActive = _isActivated(g) && !_isExpired(g);
+
+            // If this wallet is active, do nothing.
+            if (!isGuardianActive) {
+                w.guardians[pos - 1].validSince = validSince.toUint40();
+                w.guardians[pos - 1].validUntil = 0;
+            }
```

Comment:
If the guardian is pending for removal `isGuardianActive` will be true, but `validUntil` is not reset to 0 in that case, so the pending removal is not cancelled, which I think is the desired outcome.

`w.guardians[pos - 1].validUntil = 0;` I think should always be done.

Reply:
please see the new impl.